framework/di/Container.php

```diff
@@ -527,7 +531,17 @@ public function resolveCallableDependencies(callable $callback, $params = [])
                 } elseif (isset(Yii::$app) && Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
                     $args[] = $obj;
                 } else {
-                    $args[] = $this->get($className);
+                    // If the argument is optional we catch not instantiable exceptions.
+                    try {
+                        $args[] = $this->get($className);
+                    } catch (NotInstantiableException $e) {
```

Comment:
Well in general I think `$container->get()` which is public should throw a proper exception in case of misconfiguration.
And no; we still want to be able to inject optional dependencies.
Also we can't say for sure calling `$container->get()` with an interface will fail since it is a very good use of DI to map interfaces to concrete implementations.